False Negative: RedundantModifierCheck

I have read check documentation: https://checkstyle.sourceforge.io/config_modifier.html#RedundantModifier
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
/var/tmp $ javac Test.java

/var/tmp $ cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="RedundantModifier"/>
    </module>
</module>

/var/tmp $ cat Test.java
package tmp;

public interface Test {
    public static interface foo { // violation for 'static', no violation for 'public'
    }

    public static enum someEnum { // violation for both
    }

    public static class someClass { // violation for both
    }

    public static record someRecord() { // violation for none, will be fixed in #11259
    }

    public static @interface someAnnInterface { // violation for none
    }
}


/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-9.3-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /var/tmp/Test.java:4:12: Redundant 'static' modifier. [RedundantModifier]
[ERROR] /var/tmp/Test.java:7:5: Redundant 'public' modifier. [RedundantModifier]
[ERROR] /var/tmp/Test.java:7:12: Redundant 'static' modifier. [RedundantModifier]
[ERROR] /var/tmp/Test.java:10:5: Redundant 'public' modifier. [RedundantModifier]
[ERROR] /var/tmp/Test.java:10:12: Redundant 'static' modifier. [RedundantModifier]
Audit done.
Checkstyle ends with 5 errors.

```
---

**Describe what you expect in detail.**
```
Starting audit...
[ERROR] /var/tmp/Test.java:4:5: Redundant 'public' modifier. [RedundantModifier]
[ERROR] /var/tmp/Test.java:4:12: Redundant 'static' modifier. [RedundantModifier]
[ERROR] /var/tmp/Test.java:7:5: Redundant 'public' modifier. [RedundantModifier]
[ERROR] /var/tmp/Test.java:7:12: Redundant 'static' modifier. [RedundantModifier]
[ERROR] /var/tmp/Test.java:10:5: Redundant 'public' modifier. [RedundantModifier]
[ERROR] /var/tmp/Test.java:10:12: Redundant 'static' modifier. [RedundantModifier]
[ERROR] /var/tmp/Test.java:13:5: Redundant 'public' modifier. [RedundantModifier]
[ERROR] /var/tmp/Test.java:13:12: Redundant 'static' modifier. [RedundantModifier]
[ERROR] /var/tmp/Test.java:16:5: Redundant 'public' modifier. [RedundantModifier]
[ERROR] /var/tmp/Test.java:16:12: Redundant 'static' modifier. [RedundantModifier]
Audit done.
Checkstyle ends with 10 errors.
```
Every member class or interface declaration in the body of an interface declaration is implicitly public and static. See [JLS](https://docs.oracle.com/javase/specs/jls/se17/html/jls-9.html#jls-9.5) for more info. Comments in code specify the current behavior.

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java
@@ -48,7 +48,7 @@ import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
  * Final modifier on methods of final and anonymous classes.
  * </li>
  * <li>
- * Inner {@code interface} declarations that are declared as {@code static}.
+ * Type declarations nested under interfaces that are declared as {@code public} or {@code static}.
  * </li>
  * <li>
  * Class constructors.
@@ -58,13 +58,11 @@ import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
  * </li>
  * </ol>
  * <p>
- * Interfaces by definition are abstract so the {@code abstract}
- * modifier on the interface is redundant.
+ * interfaces by definition are abstract so the {@code abstract} modifier is redundant on them.
  * </p>
- * <p>Classes inside of interfaces by definition are public and static,
- * so the {@code public} and {@code static} modifiers
- * on the inner classes are redundant. On the other hand, classes
- * inside of interfaces can be abstract or non abstract.
+ * <p>Type declarations nested under interfaces by definition are public and static,
+ * so the {@code public} and {@code static} modifiers on nested type declarations are redundant.
+ * On the other hand, classes inside of interfaces can be abstract or non abstract.
  * So, {@code abstract} modifier is allowed.
  * </p>
  * <p>Fields in interfaces and annotations are automatically
@@ -155,7 +153,9 @@ import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
  * <a href="https://checkstyle.org/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#ENUM_DEF">
  * ENUM_DEF</a>,
  * <a href="https://checkstyle.org/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#RESOURCE">
- * RESOURCE</a>.
+ * RESOURCE</a>,
+ * <a href="https://checkstyle.org/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#ANNOTATION_DEF">
+ * ANNOTATION_DEF</a>.
  * </li>
  * </ul>
  * <p>
@@ -225,36 +225,53 @@ public class RedundantModifierCheck
             TokenTypes.CLASS_DEF,
             TokenTypes.ENUM_DEF,
             TokenTypes.RESOURCE,
+            TokenTypes.ANNOTATION_DEF,
         };
     }
 
     @Override
     public void visitToken(DetailAST ast) {
-        if (ast.getType() == TokenTypes.INTERFACE_DEF) {
-            checkInterfaceModifiers(ast);
+        switch (ast.getType()) {
+            case TokenTypes.INTERFACE_DEF:
+            case TokenTypes.ANNOTATION_DEF:
+                checkInterfaceModifiers(ast);
+                break;
+            case TokenTypes.ENUM_DEF:
+                checkForRedundantModifier(ast, TokenTypes.LITERAL_STATIC);
+                break;
+            case TokenTypes.CTOR_DEF:
+                checkConstructorModifiers(ast);
+                break;
+            case TokenTypes.METHOD_DEF:
+                processMethods(ast);
+                break;
+            case TokenTypes.RESOURCE:
+                processResources(ast);
+                break;
+            case TokenTypes.CLASS_DEF:
+            case TokenTypes.VARIABLE_DEF:
+            case TokenTypes.ANNOTATION_FIELD_DEF:
+                break;
+            default:
+                throw new IllegalStateException("Unexpected token type: " + ast.getType());
         }
-        else if (ast.getType() == TokenTypes.ENUM_DEF) {
-            checkEnumDef(ast);
+
+        if (isInterfaceOrAnnotationMember(ast)) {
+            processInterfaceOrAnnotation(ast);
+        }
+    }
+
+    /**
+     * Check modifiers of constructor.
+     *
+     * @param ctorDefAst ast node of type {@link TokenTypes#CTOR_DEF}
+     */
+    private void checkConstructorModifiers(DetailAST ctorDefAst) {
+        if (isEnumMember(ctorDefAst)) {
+            checkEnumConstructorModifiers(ctorDefAst);
         }
         else {
-            if (ast.getType() == TokenTypes.CTOR_DEF) {
-                if (isEnumMember(ast)) {
-                    checkEnumConstructorModifiers(ast);
-                }
-                else {
-                    checkClassConstructorModifiers(ast);
-                }
-            }
-            else if (ast.getType() == TokenTypes.METHOD_DEF) {
-                processMethods(ast);
-            }
-            else if (ast.getType() == TokenTypes.RESOURCE) {
-                processResources(ast);
-            }
-
-            if (isInterfaceOrAnnotationMember(ast)) {
-                processInterfaceOrAnnotation(ast);
-            }
+            checkClassConstructorModifiers(ctorDefAst);
         }
     }
 
@@ -286,20 +303,6 @@ public class RedundantModifierCheck
         TokenUtil.findFirstTokenByPredicate(
             modifiers, mod -> mod.getType() != TokenTypes.ANNOTATION
         ).ifPresent(modifier -> log(modifier, MSG_KEY, modifier.getText()));
-    }
-
-    /**
-     * Checks whether enum has proper modifiers.
-     *
-     * @param ast enum definition.
-     */
-    private void checkEnumDef(DetailAST ast) {
-        if (isInterfaceOrAnnotationMember(ast)) {
-            processInterfaceOrAnnotation(ast);
-        }
-        else {
-            checkForRedundantModifier(ast, TokenTypes.LITERAL_STATIC);
-        }
     }
 
     /**

--- a/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/modifier/RedundantModifierCheck.xml
+++ b/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/modifier/RedundantModifierCheck.xml
@@ -21,7 +21,7 @@
  Final modifier on methods of final and anonymous classes.
  &lt;/li&gt;
  &lt;li&gt;
- Inner {@code interface} declarations that are declared as {@code static}.
+ Type declarations nested under interfaces that are declared as {@code public} or {@code static}.
  &lt;/li&gt;
  &lt;li&gt;
  Class constructors.
@@ -31,13 +31,11 @@
  &lt;/li&gt;
  &lt;/ol&gt;
  &lt;p&gt;
- Interfaces by definition are abstract so the {@code abstract}
- modifier on the interface is redundant.
+ interfaces by definition are abstract so the {@code abstract} modifier is redundant on them.
  &lt;/p&gt;
- &lt;p&gt;Classes inside of interfaces by definition are public and static,
- so the {@code public} and {@code static} modifiers
- on the inner classes are redundant. On the other hand, classes
- inside of interfaces can be abstract or non abstract.
+ &lt;p&gt;Type declarations nested under interfaces by definition are public and static,
+ so the {@code public} and {@code static} modifiers on nested type declarations are redundant.
+ On the other hand, classes inside of interfaces can be abstract or non abstract.
  So, {@code abstract} modifier is allowed.
  &lt;/p&gt;
  &lt;p&gt;Fields in interfaces and annotations are automatically
@@ -108,7 +106,7 @@
  }
  &lt;/pre&gt;</description>
          <properties>
-            <property default-value="METHOD_DEF,VARIABLE_DEF,ANNOTATION_FIELD_DEF,INTERFACE_DEF,CTOR_DEF,CLASS_DEF,ENUM_DEF,RESOURCE"
+            <property default-value="METHOD_DEF,VARIABLE_DEF,ANNOTATION_FIELD_DEF,INTERFACE_DEF,CTOR_DEF,CLASS_DEF,ENUM_DEF,RESOURCE,ANNOTATION_DEF"
                       name="tokens"
                       type="java.lang.String[]"
                       validation-type="tokenSet">

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheckTest.java
@@ -21,10 +21,12 @@ package com.puppycrawl.tools.checkstyle.checks.modifier;
 
 import static com.google.common.truth.Truth.assertWithMessage;
 import static com.puppycrawl.tools.checkstyle.checks.modifier.RedundantModifierCheck.MSG_KEY;
+import static org.junit.jupiter.api.Assertions.assertThrows;
 
 import org.junit.jupiter.api.Test;
 
 import com.puppycrawl.tools.checkstyle.AbstractModuleTestSupport;
+import com.puppycrawl.tools.checkstyle.DetailAstImpl;
 import com.puppycrawl.tools.checkstyle.api.TokenTypes;
 import com.puppycrawl.tools.checkstyle.utils.CommonUtil;
 
@@ -126,6 +128,7 @@ public class RedundantModifierCheckTest
 
         final String[] expected = {
             "22:17: " + getCheckMessage(MSG_KEY, "public"),
+            "24:13: " + getCheckMessage(MSG_KEY, "public"),
             "26:21: " + getCheckMessage(MSG_KEY, "public"),
             "37:12: " + getCheckMessage(MSG_KEY, "public"),
             "45:17: " + getCheckMessage(MSG_KEY, "public"),
@@ -149,10 +152,26 @@ public class RedundantModifierCheckTest
             TokenTypes.CLASS_DEF,
             TokenTypes.ENUM_DEF,
             TokenTypes.RESOURCE,
+            TokenTypes.ANNOTATION_DEF,
         };
         assertWithMessage("Invalid acceptable tokens")
             .that(actual)
             .isEqualTo(expected);
+    }
+
+    @Test
+    public void testWrongTokenType() {
+        final RedundantModifierCheck obj = new RedundantModifierCheck();
+        final DetailAstImpl ast = new DetailAstImpl();
+        ast.initialize(TokenTypes.LITERAL_NULL, "null");
+
+        final IllegalStateException exception = assertThrows(IllegalStateException.class, () -> {
+            obj.visitToken(ast);
+        }, "IllegalStateException was expected");
+
+        assertWithMessage("Expected and actual violation messages do not match")
+                .that(exception.getMessage())
+                .isEqualTo("Unexpected token type: " + ast.getType());
     }
 
     @Test
@@ -273,6 +292,22 @@ public class RedundantModifierCheckTest
             "13:5: " + getCheckMessage(MSG_KEY, "static"),
             "13:12: " + getCheckMessage(MSG_KEY, "public"),
             "16:9: " + getCheckMessage(MSG_KEY, "public"),
+            "19:5: " + getCheckMessage(MSG_KEY, "public"),
+            "19:12: " + getCheckMessage(MSG_KEY, "static"),
+            "22:5: " + getCheckMessage(MSG_KEY, "public"),
+            "22:12: " + getCheckMessage(MSG_KEY, "abstract"),
+            "22:21: " + getCheckMessage(MSG_KEY, "static"),
+            "26:1: " + getCheckMessage(MSG_KEY, "abstract"),
+            "28:5: " + getCheckMessage(MSG_KEY, "public"),
+            "28:12: " + getCheckMessage(MSG_KEY, "static"),
+            "32:9: " + getCheckMessage(MSG_KEY, "public"),
+            "32:16: " + getCheckMessage(MSG_KEY, "static"),
+            "34:13: " + getCheckMessage(MSG_KEY, "public"),
+            "34:20: " + getCheckMessage(MSG_KEY, "static"),
+            "37:13: " + getCheckMessage(MSG_KEY, "public"),
+            "37:20: " + getCheckMessage(MSG_KEY, "static"),
+            "40:13: " + getCheckMessage(MSG_KEY, "public"),
+            "40:20: " + getCheckMessage(MSG_KEY, "static"),
         };
         verifyWithInlineConfigParser(getPath(
                 "InputRedundantModifierNestedDef.java"), expected);

--- a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/modifier/redundantmodifier/InputRedundantModifierNestedClassInPublicInterfaceRedundantModifiers.java
+++ b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/modifier/redundantmodifier/InputRedundantModifierNestedClassInPublicInterfaceRedundantModifiers.java
@@ -21,7 +21,7 @@ public interface InputRedundantModifierNestedClassInPublicInterfaceRedundantModi
             class ClassInPrivateNestedInterface {
                 public ClassInPrivateNestedInterface() { } // violation
             }
-            public interface PrivateNestedInterfaceWithPublicModifier {
+            public interface PrivateNestedInterfaceWithPublicModifier { // violation
                 class ClassInPrivateNestedInterface {
                     public ClassInPrivateNestedInterface() { } // violation
                 }

--- a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/modifier/redundantmodifier/InputRedundantModifierNestedDef.java
+++ b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/modifier/redundantmodifier/InputRedundantModifierNestedDef.java
@@ -15,4 +15,30 @@ public interface InputRedundantModifierNestedDef {
     interface MyInnerInterface {
         public strictfp class MyInnerClass {} // violation
     }
+
+    public static class testClass { // 2 violations
+    }
+
+    public abstract static @interface testAnnotatedInterface { // 3 violations
+    }
+}
+
+abstract @interface testAnnotatedInterface { // violation
+
+    public static enum testEnum { // 2 violations
+    }
+
+    interface testInterface {
+        public static interface nestedInterface { // 2 violations
+
+            public static class nestedClass { // 2 violations
+            }
+
+            public static @interface nestedAnnInterface { // 2 violations
+            }
+
+            public static enum nestedEnum { // 2 violations
+            }
+        }
+    }
 }

--- a/src/xdocs/config_modifier.xml
+++ b/src/xdocs/config_modifier.xml
@@ -548,8 +548,8 @@ public interface RoadFeature {
           <li>Interface and annotation definitions.</li>
           <li>Final modifier on methods of final and anonymous classes.</li>
           <li>
-            Inner <code>interface</code> declarations that are declared
-            as <code>static</code>.
+            Type declarations nested under interfaces that are declared as <code>public</code>
+            or <code>static</code>.
           </li>
           <li>Class constructors.</li>
           <li>
@@ -559,16 +559,15 @@ public interface RoadFeature {
         </ol>
 
         <p>
-          Interfaces by definition are abstract so the <code>abstract</code>
-          modifier on the interface is redundant.
+          interfaces by definition are abstract so the <code>abstract</code> modifier is
+          redundant on them.
         </p>
 
         <p>
-          Classes inside of interfaces by definition are public and static,
-          so the <code>public</code> and <code>static</code> modifiers
-          on the inner classes are redundant. On the other hand, classes
-          inside of interfaces can be abstract or non abstract.
-          So, <code>abstract</code> modifier is allowed.
+          Type declarations nested under interfaces by definition are public and static,
+          so the <code>public</code> and <code>static</code> modifiers on nested type
+          declarations are redundant. On the other hand, classes inside of interfaces can
+          be abstract or non abstract. So, <code>abstract</code> modifier is allowed.
         </p>
 
         <p>
@@ -689,6 +688,8 @@ public class ClassExtending extends ClassExample {
                   ENUM_DEF</a>
                 , <a href="apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#RESOURCE">
                   RESOURCE</a>
+                , <a href="apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#ANNOTATION_DEF">
+                ANNOTATION_DEF</a>
                 .
               </td>
               <td>
@@ -708,6 +709,8 @@ public class ClassExtending extends ClassExample {
                   ENUM_DEF</a>
                 , <a href="apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#RESOURCE">
                   RESOURCE</a>
+                , <a href="apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#ANNOTATION_DEF">
+                ANNOTATION_DEF</a>
                 .
               </td>
               <td>3.0</td>

